Add AuthnContextClassRef attribute blacklist filter

pablothedude opened the pull request:
Add a configurable regex filter to be able to blacklist certain AuthnContextClassRef attribute values. This will be used to prevent the use of internally used attributes by IdP's other than EngineBlock. And to allow second factor only (SFO) flow in EB.

Also created a [PR](https://github.com/OpenConext/OpenConext-deploy/pull/237) to support the configuration parameter in deploy.

See: https://www.pivotaltracker.com/story/show/166702153

pablothedude commented:
@domgon would you be so kind to validate the Portuguese translations in this PR?
https://github.com/OpenConext/OpenConext-engineblock/pull/719/files#diff-a9d09e45c9163e92df703f169a1c2309

domgon commented:
Hi,

the translations are bellow.

```
    'error_authn_context_class_ref_blacklisted'                     => 'Erro - O valor para AuthnContextClassRef não é permitido',
    'error_authn_context_class_ref_blacklisted_desc'                => '<p>Não pode autenticar-se porque a sua %organisationNoun% enviou um valor para AuthnContextClassRef que não é permitido.</p>',
```

Best regards,
Domingos

MKodde commented on languages/messages.nl.php:
```diff
@@ -244,6 +244,8 @@
         Je bent succesvol ingelogd bij je Identity Provider maar de dienst waar je naartoe wilt stuurt je weer terug naar %suiteName%. Omdat je succesvol bent ingelogd, stuurt %suiteName% je weer naar de dienst, wat resulteert in een oneindig zwart gat. Dit komt waarschijnlijk door een foutje aan de kant van de dienst.
     </p>',
+    'error_authn_context_class_ref_blacklisted'                     => 'Error - Waarde van AuthnContextClassRef is niet toegestaan',
+    'error_authn_context_class_ref_blacklisted_desc'                => '<p>Je kunt niet inloggen omdat je %organisationNoun% een waarde stuurde voor AuthnContextClassRef die niet is toegestaan</p>',
```
Missing '.' ?

MKodde commented on library/EngineBlock/Corto/Filter/Command/ValidateAuthnContextClassRef.php:
```diff
@@ -0,0 +1,63 @@
+use Psr\Log\LoggerInterface;
+
+/**
+ * Validate if the AuthnContextClassRef is not blacklisted.
```
This would be an excellent place to state/explain the business rules that resulted in the creation of this validator.

pablothedude replied:
If you look into git you could see why it is added.

MKodde replied:
What's wrong with adding some details about the validator to the class' doc block? Reading git log will probably get you there as well, but is far less user-friendly.

Suggestion: 

> The validator will block any incoming IdP assertion with a AuthnContextClassRef value in our namespace with an error. As it MUST be impossible for an IdP to 'impersonate' our values used for AuthnContextClassRef.

MKodde commented on library/EngineBlock/Corto/Filter/Command/ValidateAuthnContextClassRef.php:
```diff
@@ -0,0 +1,63 @@
+     */
+    private $authnContextClassRefBlacklistPattern;
+
+    public function __construct(LoggerInterface $logger, $authnContextClassRefBlacklistPattern)
```
Have you considered creating a modern day Symfony service that is injected into this legacy command? 

This way we venture away from creating new Corto BL. And increase testability. I would greatly applaud that being changed here!

pablothedude replied:
I've considered it but I didn't found it an improvement. I think it will be better to bring Corto up to date at once by fixing the issues below so it won't be considered legacy anymore:
* add namespaces
* prevent the use of superglobals
* eliminate the magic loading of services
* prevent magic method calls `$object->$parameterName()`
* introduce DI if not used

MKodde replied:
I do agree with you, you would introduce some coupling to the legacy code. Which in my opinion is not that big of a deal. Delving new BL into a legacy system in my opinion is far worse. Having the code in the new place prevents us from having to refactor it when the big bang happens. For now the logic is quite simple and the problem is not that evident. But in upcoming PR's we will be adding additional rules to this Command.

pablothedude replied:
I don't see the introduced code as legacy as the only thing which should be addressed is the lack of namespaces. If I did have encountered other smells I would have directly boyscouted it so it could be refactored very easy later on.

MKodde commented on library/EngineBlock/Corto/Filter/Command/ValidateAuthnContextClassRef.php:
```diff
@@ -0,0 +1,63 @@
+    public function execute()
+    {
+        if (empty($this->authnContextClassRefBlacklistPattern)) {
+            $this->logger->notice('No authn_context_class_ref_blacklist_regex found in the configuration, not verifying AuthnContextClassRef');
```
For the sake of consistency: ~~verifying~~ validating

MKodde replied:
Ask yourself, when reading this log message, is it directly clear why this validation is skipped? Maybe you could add something about not testing for SFO impersonation by remote IdP? I'm not fully sold on the current log message. @thijskh @pmeulen @quartje what are your thoughts on this? As you guys will be actually reading this message in the future.

pablothedude replied:
@MKodde  changed the message as suggested

MKodde commented on tests/library/EngineBlock/Test/Corto/Filter/Command/ValidateAuthnContextClassRefTest.php:
```diff
@@ -0,0 +1,78 @@
+        $verifier->execute();
+    }
+
+    public function testMatchedBlacklistedRegexpThrowsException()
```
It would be nice to make this test match more closely with the actual use case we are going to use this for. This test also advocates the move of the logic to modern EB! as per my previous comment :top: 

Idea: create a data provider that tries different AuthnContextClassRefs with different of our configured loa's.

pablothedude replied:
The only thing I can think of why the code isn't that modern, is the lack of using namespaces so I don't agree on this one.

The responsibility of the class is to match against a regex and that gets tested, the regex itself is language specific and I don't think that needs to be tested. Could you elaborate what the added benefit will be to test that logic?

MKodde replied:
I think a perfect unit test covers all of the different possible uses of a unit of code regardless of implementation. If a requirement states that a validator should accept loa-1, loa-2, loa-3 but NOT loa-4, then it makes sense to add a unit test for those rules. How the actual piece of code is written is beside the point. It's all about creating a set of tests that validate our business rules.

And for my previous comment:

> It would be nice to make this test match more closely with the actual use case we are going to use this for

What I meant by this is that instead of setting the blacklisted regex to something not to be used in real life: `/urn:oasis:names:tc:SAML:2\.0:ac:classes:Password/`, you could write the test to more closely match reality.

pablothedude replied:
I think a unittest should only test the possibilities of a certain unit. The class should filter the `AuthnContextClassRef` attribute and that gets tested.

MKodde commented on bin/composer/dump-required-ini-params.sh:
```diff
@@ -236,7 +236,10 @@ $ymlContent = array(
             )
         ),
-        'monitor_database_health_check_query'                     => $config->get('openconext.monitor_bundle_health_query', '')
+        'monitor_database_health_check_query'                     => $config->get('openconext.monitor_bundle_health_query', ''),
```
Do we want/need additional input validation? If not for our own safety, it could also help users with setting a correct regex.

pablothedude replied:
Input validation for a regex? How do you picture that?

MKodde replied:
You could test for it being either an empty string, or it being an actual string value. There are ways to validate if the provided string would function as a regular expression. But validating that seems less useful.

pablothedude replied:
A good suggestion, done

MKodde commented on library/EngineBlock/Corto/Filter/Command/ValidateAuthnContextClassRef.php:
```diff
@@ -0,0 +1,63 @@
+        if (!$this->validateAuthnContextClassRef($authnContextClassRef, $this->authnContextClassRefBlacklistPattern)) {
+            throw new EngineBlock_Corto_Exception_AuthnContextClassRefBlacklisted(
+                sprintf(
+                    'Disallowed AuthnContextClassRef value "%s" in response from IdP "%s" to SP "%s"',
```
I think we could improve this error / log message. 

Suggestion, follow the story provided message: 

>"Assertion from IdP <IdPEntityID> contains a blacklisted AuthnContextClassRef: '<the blacklisted AuthnContextClassRef>'"

Or something along the lines of:

> The IdP: "%s" provided a blacklisted AuthnContextClassRef: "%s" while sending an assertion to SP: "%s"

pablothedude replied:
I did reuse the the battle tested message from `library/EngineBlock/Corto/Filter/Command/ValidateAllowedConnection.php`. I however did change it to your first suggestion in terms of efficiency.

MKodde commented on src/OpenConext/EngineBlockBundle/Controller/FeedbackController.php:
```diff
@@ -194,6 +195,17 @@ public function missingRequiredFieldsAction()
+    /**
+     * @return Response
+     */
+    public function authnContextClassRefBlacklistedAction()
```
Didn't the story explicitly state that no additional error screen was required? 

>The error when a assertion in blocked due to a blacklisted AuthnContextClassRef does not need its own screen, but should be clearly logged. E.g. "Assertion from IdP <IdPEntityID> contains a blacklisted AuthnContextClassRef: '<the blacklisted AuthnContextClassRef>'"

MKodde replied:
Is 403 the correct error code? 400 - Forbidden also seems appropriate. Or as per the original intent, a 500 error?

pablothedude replied:
I see it, I think I misread it. Status code 400 is bad request and 403 is forbidden, what do you mean by `400 forbidden`?

MKodde replied:
My bad, 400 indeed is a bad request. Which could also apply here dont you think?

pablothedude replied:
In the ticket I've asked Thijs. For now I'll keep it a 403 and will hotfix it if another status code is required.

thijskh replied:
403 is fine I think

pablothedude replied:
:+1:

MKodde commented on src/OpenConext/EngineBlockBundle/Resources/config/routing/feedback.yml:
```diff
@@ -48,6 +48,11 @@ authentication_feedback_missing_required_fields:
     methods:    [GET]
     defaults:   { _controller: engineblock.controller.authentication.feedback:missingRequiredFieldsAction }
 
+authentication_authn_context_class_ref_blacklisted:
```
If we are keeping this error page, please also add it to the visual regression test

pablothedude replied:
Good suggestion, done

MKodde commented on src/OpenConext/EngineBlockFunctionalTestingBundle/Features/ClearErrorMessages.feature:
```diff
@@ -320,6 +320,15 @@ Feature:
     Then I should not see the same request id on the error page
 
+  Scenario: I log in at my Identity Provider, that has an invalid AuthnContextClassRef attribute.
+    Given the IdP "Dummy Idp" sends AuthnContextClassRef "invalid-authn-context-class-ref"
```
Consider setting this with more a more in context value. Like `https://engine.example.org/loa1` or whatnot. This [sh|c]ould make the test more recognizable in the future.

pablothedude replied:
I deliberately wanted to make it more unique because the regex was injected with the DiContainer and therefore it could be found much easier and prevent a certain level of WTF.
`\EngineBlock_Application_FunctionalTestDiContainer::getAuthnContextClassRefBlacklistRegex`

MKodde commented on src/OpenConext/EngineBlockFunctionalTestingBundle/Features/Context/EngineBlockContext.php:
```diff
@@ -6,6 +6,7 @@
 use Behat\Mink\Exception\ExpectationException;
 use DOMDocument;
 use DOMXPath;
+use EngineBlock_ApplicationSingleton;
```
The imports in this class can be removed I suppose?

pablothedude replied:
:+1:

MKodde commented on src/OpenConext/EngineBlockFunctionalTestingBundle/Features/ClearErrorMessages.feature:
```diff
@@ -320,6 +320,15 @@ Feature:
     Then I should not see the same request id on the error page
 
+  Scenario: I log in at my Identity Provider, that has an invalid AuthnContextClassRef attribute.
+    Given the IdP "Dummy Idp" sends AuthnContextClassRef "invalid-authn-context-class-ref"
```
For consistency with the other step definitions I suggest::
> Given the IdP "Dummy Idp" sends AuthnContextClassRef **with value** "invalid-authn-context-class-ref"

pablothedude replied:
Done

pablothedude commented on application/configs/application.ini:
```diff
@@ -161,6 +161,11 @@ pdp.username = "pdp_admin"
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;;;;;;;;;;; SFO SETTINGS ;;;;;;;;;;;;;;;;;;;;;;
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+sfo.authn_context_class_ref_blacklist_regex = ""
```
@MKodde I didn't want to focus on any documentation before a functional review was done. Your suggestions however are very valuable. I shall create a story to do this afterwards and will add the raised concerns to the story if you agree that will be a workable solution.

I did also think about about using a custom feature flag to enable/disable SFO completely so you don't need a regex to implicitly enable/disable SFO.

pablothedude replied:
Created story: https://www.pivotaltracker.com/story/show/167320950

MKodde replied:
Okay, thanks for the story! I still think this is the time to add in-line documentation to some parts of the code. For example to the application.ini. The additional wiki page is still a great idea, but I feel we should also be helpfull in a sense of in-line documentation when it comes to writing user-facing code (like config files).

pablothedude replied:
Added additional documentation to the `applicatio.ini`

pablothedude commented:
> Nice work! I've found some points of discussion during my code review, I've added them below.
> 
> Other concerns:
> 
>     1. Consider updating the readme with a section on SFO configuration. Or link to a OC wiki page where more information about this topic can be read. And improve in-line documentation (see some suggestions in code review)
> 
Created story

>     2. Are full PCRE regexes allowed? And should the users provide delimiters, or are they pre-configured? Again information that should become clear when opening the configuration.
> 
Added a bit of documentation to `application.ini` and a task to the documentation story.

>     3. Please document the additional requirement of the custom error page in Pivotal, or remove it from the PR. I personally opt for the first.
I've updated the task to reflect this.
